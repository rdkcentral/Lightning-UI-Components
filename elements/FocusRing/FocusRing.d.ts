import type { Base } from '../../Base';

declare class FocusRingStyles {
  radius?: number;
  borderWidth?: number;
  color?: number;
  secondaryColor?: number;
  transitionColor?: number;
  colorTransitionAlpha?: number;
  shouldAnimate?: boolean;
  spacing?: number;
  animationDuration?: number;
}

export declare class FocusRing extends Base {
  startAnimation(): void;
  stopAnimation(): void;
  style: FocusRingStyles;
}
