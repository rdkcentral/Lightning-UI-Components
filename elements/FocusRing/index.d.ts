import Base from '../../Base';
import { WithThemeStylesConstructor } from '../../mixins/withThemeStyles';
import type { Tools } from '@lightningjs/core';

export interface FocusRingStyles {
  radius?: Tools.CornerRadius;
  borderWidth?: number;
  color?: number;
  secondaryColor?: number;
  transitionColor?: number;
  colorTransitionAlpha?: number;
  shouldAnimate?: boolean;
  spacing?: number;
  animationDuration?: number;
}

declare const FocusRing_base: WithThemeStylesConstructor<
  typeof Base,
  FocusRingStyles
>;

export default class FocusRing extends FocusRing_base {
  startAnimation(): void;
  stopAnimation(): void;
}
