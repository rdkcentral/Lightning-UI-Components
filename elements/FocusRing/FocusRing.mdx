import { Canvas, Story } from '@storybook/addon-docs';
import { Theme } from '../../.storybook/utils';
import FocusRing from '.';

# FocusRing

Fancy ring to focus on Items

## Usage

Add a focus ring to tile or other items to provide a nice highlight
and shadow effect

```js
import { FocusRing } from '@lightning/ui';

FocusRing: {
  type: FocusRing,
  w: this.w,
  h: this.h,
  size: 12
}
```

<Canvas>
  <Story id="elements-focusring--basic" />
</Canvas>

### Properties

name|type|readonly|default|description
--|--|--|--|--
radius|number|false|Xfinity Theme medium|set radius of corners
color|color|false|Xfintity Theme palette.focusRing.primary|main color of focus ring (.96 alpha)
secondaryColor|color|false|false|Xfintity Theme palette.focusRing.secondary|secondary color of focus ring (.08 alpha)
spacing|number|false|12|sets the amount to increase the focus ring by from the initial component size

### Methods

#### startAnimation(): void

Start the animation of the border

#### stopAnimation(): void

Stops the border animation fallback is a solid border gradient of color/secondaryColor or Xfinity Theme default

### Styles

Styled components can be adjusted by composing using [withStyles](?path=/docs/mixins-withstyles--basic). Below is the default `styles` object

<Theme theme={FocusRing.styles} />
