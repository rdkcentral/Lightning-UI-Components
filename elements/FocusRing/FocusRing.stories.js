import lng from '@lightningjs/core';
import FocusRing from '.';
import mdx from './FocusRing.mdx';
import kabob from '../../assets/images/kabob_320x180.jpg';

export default {
  args: {
    animate: true
  },
  title: 'Elements / FocusRing',
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Basic = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        x: 60,
        y: 60,
        w: 320,
        h: 180,
        FocusRing: {
          type: FocusRing,
          w: 320,
          h: 180,
          radius: args.radius,
          color: args.color,
          secondaryColor: args.secondaryColor
        }
      };
    }

    _init() {
      if (args.animate) {
        this._FocusRing.startAnimation();
      } else {
        this._FocusRing.stopAnimation();
      }
    }

    get _FocusRing() {
      return this.tag('FocusRing');
    }
  };

Basic.argTypes = {
  animate: { control: 'boolean' }
};
Basic.parameters = {
  argActions: {
    animate: (play, component) => {
      if (play) {
        component._FocusRing.startAnimation();
      } else {
        component._FocusRing.stopAnimation();
      }
      component._refocus();
    }
  }
};

export const WithImage = () =>
  class Basic extends lng.Component {
    static _template() {
      return {
        x: 60,
        y: 60,
        Kabob: {
          src: kabob,
          w: 320,
          h: 180,
          zIndex: 2,
          radius: 8
        },
        FocusRing: {
          type: class extends FocusRing {
            _update() {
              super._update();
              this.startAnimation();
            }
          },
          w: 320,
          h: 180
        }
      };
    }
  };
