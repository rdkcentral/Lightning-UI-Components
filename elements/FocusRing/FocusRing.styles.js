export const base = theme => ({
  animationDuration: theme.animations.durations.xslow * 4,
  borderWidth: theme.stroke.md,
  colorTransitionAlpha: theme.alphas.secondary,
  radius: theme.radius.md,
  spacing: theme.spacer.xs,
  shouldAnimate: true
});

export const palette = theme => ({
  neutral: {
    color: theme.colors.coreNeutral,
    transitionColor: null,
    secondaryColor: theme.colors.coreNeutralTertiary
  },
  inverse: {
    color: theme.colors.coreInverse,
    transitionColor: null,
    secondaryColor: theme.colors.coreInverseTertiary
  },
  brand: {
    color: theme.colors.coreBrand,
    transitionColor: null,
    secondaryColor: theme.colors.coreBrandTertiary
  }
});
