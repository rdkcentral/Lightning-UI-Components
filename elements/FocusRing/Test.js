import lng from '@lightningjs/core';
export default class FunkyFocusRing extends lng.Component {
  static _template() {
    return {
      Test: {
        rect: true,
        w: 400,
        h: 500,
        color: 0xfff663399
      }
    };
  }
}
