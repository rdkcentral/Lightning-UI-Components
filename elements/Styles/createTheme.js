/**
 * Copyright 2020 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import lng from '@lightningjs/core';
import { clone } from '../../utils';
import { CORNER_RADIUS } from './Styles';
import { COLORS_NEUTRAL, getHexColor } from './Colors';

function glow({
  w,
  h,
  color = 0xffffffff,
  borderRadius = CORNER_RADIUS.small,
  blur = spacing(2)
}) {
  return {
    color: color,
    mount: 0.5,
    x: w / 2,
    y: h / 2 + blur / 2,
    zIndex: -1,
    texture: lng.Tools.getShadowRect(
      w - spacing(2),
      h - spacing(2),
      borderRadius,
      blur
    )
  };
}

function luminance({ w, h, blur = 3, padding = spacing(8), texture = null }) {
  return {
    type: lng.components.FastBlurComponent,
    x: w / 2,
    y: h / 2 + spacing(3),
    mount: 0.5,
    w,
    h,
    amount: blur,
    padding: padding,
    content: {
      Image: {
        w,
        h,
        texture: texture
      }
    },
    zIndex: -1
  };
}

const materials = {
  glow,
  luminance
};

function spacing(multiplier) {
  return 8 * multiplier;
}

const gradientColor = COLORS_NEUTRAL.light2;
const gradientAnimation = {
  duration: 0.6,
  actions: [
    {
      p: 'colorUl',
      v: {
        0: getHexColor(gradientColor, 72),
        1: getHexColor(gradientColor, 56)
      }
    },
    {
      p: 'colorUr',
      v: {
        0: getHexColor(gradientColor, 24),
        1: getHexColor(gradientColor, 16)
      }
    },
    {
      p: 'colorBr',
      v: { 0: 0x00, 1: getHexColor(gradientColor, 0) }
    },
    {
      p: 'colorBl',
      v: {
        0: getHexColor(gradientColor, 24),
        1: getHexColor(gradientColor, 16)
      }
    }
  ]
};

const SIZES = {
  icon: {
    small: 32,
    medium: 40,
    large: 48
  }
};

/**
 * Combines two theme objects to create a new theme
 * @param {Object|function} theme - An object or callback that accepts the default theme as an argument. Gets merged with the default theme
 * @param {Object} [defaultTheme] - Theme to be merged with.
 */
export default (theme, defaultTheme = {}) => {
  if (typeof theme === 'function') {
    return clone(defaultTheme, theme(defaultTheme));
  }

  return clone(defaultTheme, theme);
};
