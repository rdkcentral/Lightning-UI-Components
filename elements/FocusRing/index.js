import lng from '@lightningjs/core';
import Base from '../../Base';
import { withExtensions } from '../../mixins';
import withStyles from '../../mixins/withThemeStyles';
import styles from './FocusRing.styles';
import { getHexColor, getValidColor } from '../../Styles/Styles';

class FocusRing extends Base {
  static get __componentName() {
    return 'FocusRing';
  }

  static get tags() {
    return ['Ring'];
  }

  _inactive() {
    this.stopAnimation();
  }

  _update() {
    this._updateRing();
    this._updateAnimation();
  }

  get _ringColors() {
    return {
      primary: this._componentStyles.color,
      transition:
        this._componentStyles.transitionColor ||
        getHexColor(
          getValidColor(this._componentStyles.color),
          this._componentStyles.colorTransitionAlpha
        ),
      secondary: this._componentStyles.secondaryColor
    };
  }

  _updateRing() {
    const { transition, primary, secondary } = this._ringColors;
    const offset =
      this._componentStyles.spacing * 2 + this._componentStyles.borderWidth;
    const focusRingPadding =
      this._componentStyles.spacing + this._componentStyles.borderWidth / 2;
    let radius = [];
    if (Array.isArray(this._componentStyles.radius)) {
      radius = new Array(4)
        .fill()
        .map(
          (_, index) =>
            (this._componentStyles.radius[index] || 0) + focusRingPadding
        );
    } else {
      radius = this._componentStyles.radius
        ? this._componentStyles.radius + focusRingPadding
        : 0;
      radius = Math.max(0, radius); // Ensure number is always positive
    }
    this.patch({
      Ring: {
        mount: 0.5,
        x: this.w / 2,
        y: this.h / 2,
        alpha: this.w && this.h ? 1 : 0.001,
        texture: lng.Tools.getRoundRect(
          this.w + offset,
          this.h + offset,
          radius,
          this._componentStyles.borderWidth,
          false,
          false,
          false
        ),
        colorUl: primary,
        colorBl: transition,
        colorUr: transition,
        colorBr: secondary
      }
    });
  }

  _updateAnimation() {
    if (!this._componentStyles.shouldAnimate) {
      this._focusRingAnimation = null;
      return;
    }

    const { transition, primary, secondary } = this._ringColors;
    if (this.stage.animations.active.has(this._focusRingAnimation)) {
      // Delete the animation so it can recieve the new update
      this.stage.animations.active.delete(this._focusRingAnimation); // TODO: Ask metrological about this approach, there should be something added to core to allow this
      this._focusRingAnimation = undefined;
    }
    this._focusRingAnimation = this._Ring.animation({
      repeat: -1,
      duration: this._componentStyles.animationDuration,
      actions: [
        {
          p: 'colorUl',
          v: {
            0: primary,
            0.25: transition,
            0.5: secondary,
            0.75: transition,
            1: primary
          }
        },
        {
          p: 'colorUr',
          v: {
            0: transition,
            0.25: primary,
            0.5: transition,
            0.75: secondary,
            1: transition
          }
        },
        {
          p: 'colorBl',
          v: {
            0: transition,
            0.25: secondary,
            0.5: transition,
            0.75: primary,
            1: transition
          }
        },
        {
          p: 'colorBr',
          v: {
            0: secondary,
            0.25: transition,
            0.5: primary,
            0.75: transition,
            1: secondary
          }
        }
      ]
    });
    if (this._isAnimating) this._focusRingAnimation.start();
  }

  startAnimation() {
    if (!this._Ring || !this._focusRingAnimation) {
      // Ring is not setup yet, store the flag to start animation once it is
      this._isAnimating = true;
      return;
    }

    if (!this._componentStyles.shouldAnimate) {
      this.stopAnimation();
      return;
    }

    this._isAnimating = true;
    this._focusRingAnimation.start();
  }

  stopAnimation() {
    if (!this._Ring || !this._focusRingAnimation) {
      // Ring is not setup yet, store the flag to make sure animation does not start when created
      this._isAnimating = false;
      return;
    }

    this._isAnimating = false;
    this._focusRingAnimation.stop();
  }
}

export default withExtensions(withStyles(FocusRing, styles));
