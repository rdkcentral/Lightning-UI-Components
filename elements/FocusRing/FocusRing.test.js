import TestUtils from '../../test/lightning-test-utils';
import FocusRing from '.';
import { getHexColor, GREY } from '../../Styles';

const createFocusRing = TestUtils.makeCreateComponent(FocusRing, {
  w: 320,
  h: 180,
  radius: 8
});

describe('FocusRing', () => {
  let focusRing, testRenderer;

  beforeEach(() => {
    [focusRing, testRenderer] = createFocusRing();
  });

  afterEach(() => {
    focusRing = null;
    testRenderer = null;
  });

  it('should render', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('should set size', () => {
    [focusRing, testRenderer] = createFocusRing({ size: 4 });
    expect(focusRing.size).toEqual(4);
  });

  it('should set primary color accounting for color type passed', () => {
    const testColor = getHexColor(GREY[5], 96);
    [focusRing, testRenderer] = createFocusRing({
      style: { color: getHexColor(GREY[5], 96) }
    });
    expect(focusRing._Ring.colorUl).toEqual(testColor);
  });

  it('should set primary color transition from color', () => {
    const testColor = getHexColor(getHexColor('#663399', 96), 60);
    [focusRing, testRenderer] = createFocusRing({
      style: {
        color: getHexColor(getHexColor('#663399', 96), 60)
      }
    });
    const { transition } = focusRing._ringColors;
    expect(transition).toEqual(testColor);
  });

  it('should set secondary color', () => {
    [focusRing, testRenderer] = createFocusRing({
      style: {
        secondaryColor: getHexColor('#663399', 8)
      }
    });
    const { secondary } = focusRing._ringColors;
    expect(secondary).toEqual(getHexColor('#663399', 8));
  });

  it('should set radius and include the size of the ring', done => {
    [focusRing, testRenderer] = createFocusRing({ style: { radius: 12 } });
    expect(
      parseInt(focusRing.tag('Ring').texture._lookupId.split(',').pop())
    ).toEqual(18);
    done();
  });

  it('should set the spacing', () => {
    const spacing = 50;
    [focusRing, testRenderer] = createFocusRing({ style: { spacing } });
    const rectSettings = focusRing
      .tag('Ring')
      .texture._lookupId.replace('rect', '')
      .split(',');
    const offset = spacing * 2 + focusRing._componentStyles.borderWidth;
    expect(parseInt(rectSettings[0])).toEqual(focusRing.w + offset);
    expect(parseInt(rectSettings[1])).toEqual(focusRing.h + offset);
  });

  it('should start and stop animating', () => {
    [focusRing, testRenderer] = createFocusRing();
    focusRing.startAnimation();
    expect(focusRing._focusRingAnimation.isPlaying()).toBe(true);
    focusRing.stopAnimation();
    expect(focusRing._focusRingAnimation.isPlaying()).toBe(false);
  });

  it('should stop animating if the component becomes inactive', () => {
    [focusRing, testRenderer] = createFocusRing();
    focusRing.startAnimation();
    expect(focusRing._focusRingAnimation.isPlaying()).toBe(true);
    focusRing._inactive();
    expect(focusRing._focusRingAnimation.isPlaying()).toBe(false);
  });

  it('should not animate if shouldAnimate is set to false in the styles', () => {
    [focusRing, testRenderer] = createFocusRing({
      style: { shouldAnimate: false }
    });
    expect(focusRing._componentStyles.shouldAnimate).toBe(false);
    focusRing.startAnimation();
    expect(focusRing._focusRingAnimation).toBe(null);
  });
});
