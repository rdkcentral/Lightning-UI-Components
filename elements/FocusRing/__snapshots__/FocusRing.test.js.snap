// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusRing should render 1`] = `
Object {
  "Component": Object {
    "Ring": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294375161,
      "colorBl": "99f6f6f9",
      "colorBr": "26f6f6f9",
      "colorUl": "fff6f6f9",
      "colorUr": "99f6f6f9",
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0.5,
      "mountX": 0.5,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Ring",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Ring",
      ],
      "texture": Object {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 160,
      "y": 90,
      "zIndex": 0,
    },
  },
}
`;
